src/main/java/com/couchbase/lite/Database.java

```diff
@@ -2194,48 +2196,55 @@ protected void rememberAttachmentWriter(BlobStoreWriter writer) {
-                changesToNotify.clear();
+            // This is a 'while' instead of an 'if' because when we finish posting notifications, there
+            // might be new ones that have arrived as a result of notification handlers making document
+            // changes of their own (the replicator manager will do this.) So we need to check again.
+            while (!store.inTransaction() && isOpen() && changesToNotify.size() > 0) {
+                try {
+                    List<DocumentChange> outgoingChanges = new ArrayList<DocumentChange>();
+                    synchronized (changesToNotify) {
+                        outgoingChanges.addAll(changesToNotify);
+                        changesToNotify.clear();
+                    }
 
-                // TODO: postPublicChangeNotification in CBLDatabase+Internal.m should replace
-                // following lines of code.
+                    // TODO: postPublicChangeNotification in CBLDatabase+Internal.m should replace
+                    // following lines of code.
 
-                boolean isExternal = false;
-                for (DocumentChange change : outgoingChanges) {
-                    Document doc = cachedDocumentWithID(change.getDocumentId());
-                    if (doc != null) {
-                        doc.revisionAdded(change, true);
-                    }
-                    if (change.getSource() != null) {
-                        isExternal = true;
+                    boolean isExternal = false;
+                    for (DocumentChange change : outgoingChanges) {
+                        Document doc = cachedDocumentWithID(change.getDocumentId());
+                        if (doc != null)
+                            doc.revisionAdded(change, true);
+                        if (change.getSource() != null)
+                            isExternal = true;
                     }
-                }
 
-                ChangeEvent changeEvent = new ChangeEvent(this, isExternal, outgoingChanges);
+                    final ChangeEvent changeEvent = new ChangeEvent(this, isExternal, outgoingChanges);
+                    synchronized (changeListeners) {
+                        for (ChangeListener changeListener : changeListeners)
+                            changeListener.changed(changeEvent);
+                    }
 
-                for (ChangeListener changeListener : changeListeners) {
-                    changeListener.changed(changeEvent);
+                    posted = true;
+                } catch (Exception e) {
+                    Log.e(Database.TAG, this + " got exception posting change notifications", e);
                 }
-
-                posted = true;
-            } catch (Exception e) {
-                Log.e(Database.TAG, this + " got exception posting change notifications", e);
-            } finally {
+            }
+            return posted;
+        } finally {
+            synchronized (lockPostingChangeNotifications) {
```

Comment:
If `postingChangeNotifications` variable is not set to `false`, `postChangeNotifications()` method can not be execute any more. Between Line 2206 and Line 2241, the method does not return, and it does not throw an Exception. From code point of view, using `finally` block is not necessary. However there is possibility methods that `postChangeNotifications()` method calls could throw RuntimeException. So this is reason, I use finally block to reset `postingChangeNotifications` variable. What do you think?